core/src/main/java/com/linecorp/armeria/server/throttling/tokenbucket/TokenBucketThrottlingStrategy.java

```diff
@@ -0,0 +1,116 @@
+import io.github.bucket4j.local.LocalBucketBuilder;
+
+/**
+ * A {@link ThrottlingStrategy} that provides a throttling strategy based on Token-Bucket algorithm.
+ * The throttling works by examining the number of requests from the beginning, and
+ * throttling if the request rate exceed the configured bucket limits.
+ */
+public class TokenBucketThrottlingStrategy<T extends Request> extends RetryThrottlingStrategy<T> {
+    private final AsyncBucket bucket;
+    private TokenBucketConfig config;
+    private String retryAfterSeconds;
+
+    /**
+     * Creates a new strategy with specified configuration and name.
+     * @param config Token-Bucket configuration
+     * @param name the name of the strategy
+     */
+    public TokenBucketThrottlingStrategy(@Nonnull TokenBucketConfig config, @Nullable String name) {
+        super(name);
+        setConfig(config);
+        // construct the bucket builder
+        final LocalBucketBuilder builder = Bucket4j.builder().withNanosecondPrecision();
+        for (BandwidthLimit limit : config.limits()) {
+            builder.addLimit(limit.bandwidth());
+        }
+        // build the bucket
+        bucket = builder.build().asAsync();
+    }
+
+    /**
+     * Creates a new strategy with specified configuration and default name.
+     * @param config Token-Bucket configuration
+     */
+    public TokenBucketThrottlingStrategy(@Nonnull TokenBucketConfig config) {
+        this(config, "TokenBucketThrottlingStrategy");
+    }
+
+    private void setConfig(@Nonnull TokenBucketConfig config) {
+        this.config = Objects.requireNonNull(config);
+        retryAfterSeconds = String.valueOf(config.retryAfterTimeout().getSeconds());
+    }
+
+    /**
+     * Returns {@link TokenBucketConfig}.
+     */
+    @Nonnull
+    public TokenBucketConfig getConfig() {
+        return config;
+    }
+
+    /**
+     * Provides a performance shortcut to {@link TokenBucketConfig#retryAfterTimeout()}.
+     */
+    @Override
+    protected String retryAfterSeconds() {
+        return retryAfterSeconds;
+    }
+
+    /**
+     * Allows resetting the Token-Bucket configuration at runtime by providing a new set of limits.
+     * Empty set of limits will remove previously set limits.
+     * @param config {@link TokenBucketConfig}
+     */
+    public CompletableFuture<Void> replaceConfiguration(@Nonnull TokenBucketConfig config) {
+        setConfig(config);
+
+        // construct the configuration builder
+        final ConfigurationBuilder builder = Bucket4j.configurationBuilder();
+        for (BandwidthLimit limit : config.limits()) {
+            builder.addLimit(limit.bandwidth());
+        }
+        // reconfigure the bucket
+        return bucket.replaceConfiguration(builder.build());
+    }
+
+    /**
+     * Registers a request with the bucket.
+     */
+    @Override
+    public CompletionStage<Boolean> accept(ServiceRequestContext ctx, T request) {
+        return bucket.tryConsume(1L);
```

Comment:
FWIU, token bucket is designed for managing connections with varying bandwidth. Can you describe the advantages of this over Guava's algorithm when using a fixed bandwidth (`1`).

Reply:
I don't have the details of Guava's algorithm. In their documentation they just say it's "simple". I would say it's too simple to match the needs of our systems. On the contrary, [Token Bucket](https://en.wikipedia.org/wiki/Token_bucket) is an established algorithm serving wide variety use cases. Token Bucket and Leaky Bucket algorithms are standard to the industry I work for (telco) primarily used for traffic shaping and congestion control. There are multiple implementations available. I've picked the most popular one for JVM.
Now, comparing the rate limiting implementations of Guava and Bucket4j. Here are some highlights:
- Precision: `Bucket4j.builder().withNanosecondPrecision()` - allows to go up to nanosecond precision.
- Refill strategy: `Refill.intervally(10, Duration.ofSeconds(1))` - it's possible to customize refill strategy. The refill can be "greedy" (asap), interval based or take place only at reaching the end of the interval.
- Multiple limits: This is invaluable to be able to control the bursts of traffic.
- Returning tokens back to bucket: `bucket.addTokens(50)` - tokens could be returned to the bucket.
- Bucket configuration can change on the fly: `bucket.replaceConfiguration(newConfiguration)` - so it's possible to implement several traffic shaping schemes, weekday vs weekend, etc.
- Bucket can be distributed when configured to use distributed cache like Apache Ignite. This way it's possible to do throttling across the cluster.
- Bucket4j provides asynchronous API: compare `return bucket.tryConsume(1L)` with return `completedFuture(rateLimiter.tryAcquire())` from `RateLimitingThrottlingStrategy`.
- Bucket4j supports monitoring via bucket events.

Reply:
Thanks for the extra links. It made me think that we might need a `BucketSelector`? It seems like this is most effective when each client has it's own bucket, to prevent one bad client from affecting others. With a global bucket, it seems less clear what a client should do if returned the rate limit information like the bitbucket example since other clients will also be using the bucket.

I guess we could add a client decorator too that attaches a client id (uuid I guess) and one bucket selector would be to use the header. Or for HTTP/2, a reasonable one would be bucket per connection. Though that seems to overlap with HTTP/2 flow control so maybe not as useful.